File: safehere_flutterapp/lib/models/journey.dart
class Journey {
  final String id;
  final String fromLabel;
  final String toLabel;
  final double? fromLat;
  final double? fromLng;
  final double? toLat;
  final double? toLng;
  final String status; // planned | active | completed | deviated
  final List<String> trustedContactIds;
  final List<String> trustedContactNames;
  final bool notifyOnDeviation;
  final double? currentLat;
  final double? currentLng;
  final int routeSafetyPercent;
  final int? etaMinutes;
  final double offRouteDistanceM;

  Journey({
    required this.id,
    required this.fromLabel,
    required this.toLabel,
    this.fromLat,
    this.fromLng,
    this.toLat,
    this.toLng,
    required this.status,
    required this.trustedContactIds,
    required this.trustedContactNames,
    required this.notifyOnDeviation,
    this.currentLat,
    this.currentLng,
    required this.routeSafetyPercent,
    this.etaMinutes,
    required this.offRouteDistanceM,
  });

  bool get isDeviated => status == 'deviated';
  bool get isActive => status == 'active' || status == 'deviated';

  factory Journey.fromJson(Map<String, dynamic> json) => Journey(
        id: json['id'],
        fromLabel: json['from_label'],
        toLabel: json['to_label'],
        fromLat: (json['from_lat'] as num?)?.toDouble(),
        fromLng: (json['from_lng'] as num?)?.toDouble(),
        toLat: (json['to_lat'] as num?)?.toDouble(),
        toLng: (json['to_lng'] as num?)?.toDouble(),
        status: json['status'],
        trustedContactIds: List<String>.from(json['trusted_contact_ids'] ?? []),
        trustedContactNames: List<String>.from(json['trusted_contact_names'] ?? []),
        notifyOnDeviation: json['notify_on_deviation'] ?? true,
        currentLat: (json['current_lat'] as num?)?.toDouble(),
        currentLng: (json['current_lng'] as num?)?.toDouble(),
        routeSafetyPercent: json['route_safety_percent'] ?? 100,
        etaMinutes: json['eta_minutes'],
        offRouteDistanceM: (json['off_route_distance_m'] as num?)?.toDouble() ?? 0.0,
      );
}

File: backend_safeher/app/models/journey.py
import uuid
from datetime import datetime, timezone
import enum

from beanie import Document
from pydantic import Field
from pymongo import IndexModel


def _uuid() -> str:
    return str(uuid.uuid4())


class JourneyStatus(str, enum.Enum):
    planned = "planned"
    active = "active"
    completed = "completed"
    deviated = "deviated"


class Journey(Document):
    id: str = Field(default_factory=_uuid)
    owner_id: str

    from_label: str
    to_label: str
    from_lat: float | None = None
    from_lng: float | None = None
    to_lat: float | None = None
    to_lng: float | None = None

    status: JourneyStatus = JourneyStatus.planned
    expected_arrival: datetime | None = None

    # Destination live-location sharing: the user can pick one or more trusted
    # contacts to share this journey with (see requirement "Destination Live
    # Location Sharing" — only these contacts receive updates for this journey,
    # separate from the emergency override which always notifies everyone).
    trusted_contact_ids: list[str] = Field(default_factory=list)
    trusted_contact_names: list[str] = Field(default_factory=list)
    notify_on_deviation: bool = True

    current_lat: float | None = None
    current_lng: float | None = None
    route_safety_percent: int = 100
    eta_minutes: int | None = None
    off_route_distance_m: float = 0.0

    # route-deviation detection state (see app/utils/geo.py)
    baseline_lat: float | None = None
    baseline_lng: float | None = None
    axis_bearing: float | None = None
    axis_locked: bool = False
    deviation_streak: int = 0

    started_at: datetime | None = None
    ended_at: datetime | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "journeys"
        indexes = [
            IndexModel("owner_id"),
        ]
